content/posts/deta_fastapi_jwt_auth_part_1.md

```diff
@@ -0,0 +1,217 @@
+        return self.hasher.verify(password, encoded_password)
+```
+
+So far, we just imported all the tools from the libraries, and we created the `Auth` class with two functions. We don't want to store the plain text password in our `users` Base. Therefore, we can use the `encode_password` function to encode the password using the `passlib['bcrypt']` library. We can store this encoded password in our `users_db` base when the user makes an account. 
+
+We also have another function `verify_password` which checks if the plain password and the encoded password from `users_db` match. This can be useful to verify user in the `/login` endpoint. 
+
+Notice that the variable `secret` is getting something from `config.py` , so let's add that key to our file.
+
+Here is a look at the `config.py` file:
+
+```python
+SECRET="sUpErSeCrEt"
+```
+
+Now that we have a way to verify passwords, and hash passwords, it is time to handle the logic for encoding and decoding JSON web tokens. The tokens are the essence of auth logic. 
+
+Inside the `Auth` class, add the following functions.  
+
+```python
+def encode_token(self, username):
+        payload = {
+            'exp' : datetime.utcnow() + timedelta(days=0, minutes=30),
+            'iat' : datetime.utcnow(),
+            'sub' : username
+        }
+        return jwt.encode(
+            payload, 
+            self.secret,
+            algorithm='HS256'
+        )
+    
+def decode_token(self, token):
+        try:
+            payload = jwt.decode(token, self.secret, algorithms=['HS256'])
+            return payload['sub']
+        except jwt.ExpiredSignatureError:
+            raise HTTPException(status_code=401, detail='Token expired')
+        except jwt.InvalidTokenError:
+            raise HTTPException(status_code=401, detail='Invalid token')
+```
+
+The `encode_token` function takes a username as a parameter and uses `pyjwt` to encode the token. We are using `timedelta` to set the expiry of the token for 30 mins. We can use this function inside the `/login` endpoint, and return a token to the client. 
+
+`decode_token` takes a token as a parameter, and attempts to decode it using the `secret`. If there are any errors like expired token or an invalid token, we can simply raise an `HTTPException`. Otherwise, we can return the username. This will be helpful to us when the client interacts with protected data, functions, etc. We can use this function to simply verify if they have access to the response. 
+
+That is all we need for the auth logic! Here is what the file looks like at the end:
+
+`auth.py`
+
+```python
+import jwt # used for encoding and decoding jwt tokens
+from fastapi import HTTPException # used to handle error handling
+from passlib.context import CryptContext # used for hashing the password 
+from datetime import datetime, timedelta # used to handle expiry time for tokens
+from config import SECRET
+
+class Auth():
+    hasher= CryptContext(schemes=['bcrypt'])
+    secret = SECRET
```

Comment:
```suggestion
```python
import os
import jwt # used for encoding and decoding jwt tokens
from fastapi import HTTPException # used to handle error handling
from passlib.context import CryptContext # used for hashing the password 
from datetime import datetime, timedelta # used to handle expiry time for tokens
from config import SECRET

class Auth():
    hasher= CryptContext(schemes=['bcrypt'])
    secret = os.getenv("APP_SECRET_STRING")
```